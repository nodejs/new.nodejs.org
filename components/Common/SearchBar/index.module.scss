.searchBarContainer {
  background-color: var(--color-fill-top-nav);
  border-radius: 6px;
  box-shadow: none;
  display: flex;
  flex-direction: column;
  max-height: 50vh;
  max-width: 450px;
  min-height: min-content;
  min-width: min-content;
  overflow-y: hidden;

  &.expanded {
    box-shadow: 0px 2px 12px 3px rgba(153, 204, 125, 0.14);
  }

  .searchInputContainer {
    align-items: center;
    display: flex;
    position: relative;
    width: 100%;

    .searchIcon,
    .closeIcon {
      color: var(--color-text-accent);
      padding: 1rem;
      vertical-align: middle;
    }

    .closeIcon {
      padding: 0 1rem 0 0;
    }

    .searchLabel {
      align-items: center;
      color: var(--color-text-accent);
      cursor: pointer;
      display: flex;
      flex: 1;
      font-weight: var(--font-weight-semibold);
      transition: all 250ms ease-in-out;
      width: 100%;
    }

    .searchInput {
      background-color: transparent;
      border: none;
      border-radius: 6px;
      color: var(--color-text-primary);
      font-size: var(--font-size-body1);
      font-weight: var(--font-weight-semibold);
      height: 100%;
      outline: none;
      width: 100%;

      &:focus {
        outline: none;
      }
    }
  }

  .searchResults {
    display: flex;
    flex-direction: column;
    height: 100%;
    min-height: 5em;
    overflow-y: auto;
    width: 100%;

    &.noResults {
      justify-content: center;
    }

    .loadingMessage {
      align-self: center;
      color: var(--color-text-primary);
      display: flex;
      font-size: 14px;
      justify-content: center;
      justify-self: center;
      white-space: nowrap;
    }
  }

  .searchResultsList {
    list-style: circle;
    margin: 0;
    padding: 0 0 0 var(--space-32);
  }

  .searchResultsFooter {
    border-top: var(--space-01) solid var(--black2);
  }

  .searchResult {
    color: var(--color-text-primary);
    display: list-item;
    margin: 0;
    padding: 3px;

    &:hover {
      font-weight: bold;
    }

    .searchResultLink {
      display: block;
      padding: 2px;
      text-decoration: none;
      text-overflow: ellipsis;

      &:focus {
        outline: var(--brand3) dotted 2px;
      }
    }

    .searchResultText {
      text-transform: capitalize;
    }

    .searchResultCode {
      border-radius: 0.3rem;
      font-weight: var(--font-weight-light);
      padding: 0 6px;
    }
  }
}

[data-theme='light'] .searchResultCode {
  background-color: var(--black2);
  color: var(--black9);
}

[data-theme='dark'] .searchResultCode {
  background-color: var(--black9);
  color: var(--black2);
}
