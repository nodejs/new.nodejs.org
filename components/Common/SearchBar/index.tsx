import { MdTravelExplore, MdClose } from 'react-icons/md';
import classNames from 'classnames';
import { AnimatePresence, motion } from 'framer-motion';
import Link from 'next/link';
import { useCallback, useEffect, useRef, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import styles from './index.module.scss';
import SectionTitle from '../SectionTitle';
import { useClickOutside } from '../../../hooks/useClickOutside';
import useKeyPress from '../../../hooks/useKeypress';
import type { FC } from 'react';
import type { SearchFunction, SearchResult } from '../../../types';

type SearchBarProps = {
  search?: SearchFunction;
  setup?: () => Promise<void>;
  footer?: FC | (() => JSX.Element);
};

const containerTransition = { type: 'spring', damping: 22, stiffness: 150 };

const containerVariants = {
  expanded: {
    width: 'auto',
    height: 'auto',
  },
  collapsed: {
    width: 0,
    height: '3em',
  },
};

const SearchBar: FC<SearchBarProps> = ({
  setup,
  search,
  footer: Footer,
}: SearchBarProps) => {
  const [isLoading, setIsLoading] = useState(true);
  const searchInputRef = useRef<HTMLInputElement | null>(null);
  const listRef = useRef<HTMLUListElement | null>(null);
  const activeResultRef = useRef<number>(-1);

  const [isExpanded, setExpanded] = useState(false);
  const [query, setQuery] = useState('');
  const [results, setResults] = useState<SearchResult[]>([]);
  const isEmpty = !results || results.length === 0;

  const containerClassNames = classNames(styles.searchBarContainer, {
    [styles.expanded]: isExpanded,
  });

  const resultsClassNames = classNames(styles.searchResults, {
    [styles.noResults]: isEmpty,
  });

  const expandContainer = useCallback(() => {
    setExpanded(true);
  }, [setExpanded]);

  const collapseContainer = useCallback(() => {
    setQuery('');
    setExpanded(false);
  }, [setQuery, setExpanded]);

  const selectItem = useCallback(
    (index: number) => {
      if (!listRef.current) {
        return;
      }

      const el = listRef.current.children[index]!;
      el.querySelector('a')?.focus();
      el.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
    },
    [listRef]
  );

  const toggleContainer = useCallback(() => {
    isExpanded ? collapseContainer() : expandContainer();
  }, [isExpanded, collapseContainer, expandContainer]);

  const onKeyPressHandler = useCallback(() => {
    if (!isExpanded) {
      expandContainer();
    }
  }, [isExpanded, expandContainer]);

  const onQueryChangeHandler = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      e.preventDefault();

      // Clear results when the query is emptied
      if (!e.target.value) {
        setResults([]);
      }

      setQuery(e.target.value);
    },
    [setQuery, setResults]
  );

  const onBlurHandler = useCallback(
    (e: React.FocusEvent<HTMLInputElement>) => {
      if (!e.currentTarget.contains(e.target) || isEmpty) {
        collapseContainer();
      }
    },
    [isEmpty, collapseContainer]
  );

  const onCloseHandler = useCallback(
    (e: React.MouseEvent<SVGSVGElement>) => {
      e.stopPropagation();
      if (isExpanded) {
        collapseContainer();
      }
    },
    [isExpanded, collapseContainer]
  );

  const onSearchResultKeyHandler = useCallback(
    (e: React.KeyboardEvent) => {
      if (!isExpanded || isEmpty || !listRef.current) {
        return;
      }

      const currentResultIndex = activeResultRef.current;
      const maxResultIndex = listRef.current.children.length - 1;

      if (e.key === 'ArrowDown') {
        e.preventDefault();

        activeResultRef.current =
          currentResultIndex + 1 > maxResultIndex ? 0 : currentResultIndex + 1;
        selectItem(activeResultRef.current);
      } else if (e.key === 'ArrowUp') {
        e.preventDefault();

        activeResultRef.current =
          currentResultIndex - 1 < 0 ? maxResultIndex : currentResultIndex - 1;
        selectItem(activeResultRef.current);
      }
    },
    [isExpanded, isEmpty, activeResultRef, selectItem]
  );

  const parentRef = useClickOutside<HTMLDivElement>(collapseContainer);

  // Focus the field when expanding the bar
  useEffect(() => {
    if (isExpanded) {
      searchInputRef.current?.focus();
    }
  }, [isExpanded, searchInputRef]);

  // Perform the setup call, if any
  useEffect(() => {
    if (!setup) {
      setIsLoading(false);
      return;
    }

    setup()
      .then(() => {
        setIsLoading(false);
      })
      .catch((error: Error) => {
        console.error('Cannot load search data.', error);
      });
  }, [setup, setIsLoading]);

  // Perform a search when the query changes
  useEffect(() => {
    if (isLoading || !query || !search) {
      return;
    }

    try {
      const searchReturnValue = search(query);

      if (
        typeof (searchReturnValue as Promise<SearchResult[]>).then ===
        'function'
      ) {
        (searchReturnValue as Promise<SearchResult[]>)
          .then(r => setResults(r))
          .catch((error: Error) => {
            console.error('Cannot perform async search.', error);
          });
      } else {
        setResults(searchReturnValue as SearchResult[]);
      }
    } catch (error) {
      console.error('Cannot perform search.', error);
    }
  }, [isLoading, query, search, setResults]);

  useKeyPress({
    targetKey: 'ctrl+k',
    callback: toggleContainer,
    preventDefault: true,
  });

  useKeyPress({
    targetKey: 'meta+k',
    callback: toggleContainer,
    preventDefault: true,
  });

  useKeyPress({
    targetKey: 'Escape',
    callback() {
      if (isExpanded) {
        collapseContainer();
      }
    },
    preventDefault: true,
  });

  return (
    <motion.div
      className={containerClassNames}
      animate={isExpanded ? 'expanded' : 'collapsed'}
      initial="collapsed"
      variants={containerVariants}
      transition={containerTransition}
      ref={parentRef}
      onBlur={onBlurHandler}
      onKeyDown={onSearchResultKeyHandler}
    >
      <div
        className={styles.searchInputContainer}
        onKeyPress={onKeyPressHandler}
        onClick={expandContainer}
        role="presentation"
      >
        <MdTravelExplore className={styles.searchIcon} />
        <label className={styles.searchLabel} htmlFor="searchInput">
          <span>
            {!isExpanded && (
              <FormattedMessage id="components.searchBar.placeholder" />
            )}
          </span>
          <input
            ref={searchInputRef}
            autoComplete="off"
            className={styles.searchInput}
            id="searchInput"
            name="query"
            type="text"
            value={query}
            onFocus={expandContainer}
            onChange={onQueryChangeHandler}
          />
        </label>
        <AnimatePresence>
          {isExpanded && (
            <motion.span
              key="close-icon"
              initial={{ opacity: 0 }}
              animate={{ opacity: 1 }}
              exit={{ opacity: 0 }}
              transition={{ duration: 0.2 }}
            >
              <MdClose className={styles.closeIcon} onClick={onCloseHandler} />
            </motion.span>
          )}
        </AnimatePresence>
      </div>

      {isExpanded && (
        <>
          <div className={resultsClassNames}>
            {isLoading && (
              <div className={styles.loadingMessage}>
                <FormattedMessage id="components.searchBar.search.loading" />
              </div>
            )}

            {!isLoading && isEmpty && (
              <div className={styles.loadingMessage}>
                <FormattedMessage
                  id={
                    query.length
                      ? 'components.searchBar.search.noResults'
                      : 'components.searchBar.search.title'
                  }
                />
              </div>
            )}
            {!isLoading && !isEmpty && (
              <ul ref={listRef} className={styles.searchResultsList}>
                {results.map((result: SearchResult) => {
                  const sectionPath =
                    result.category === 'api'
                      ? ['home', result.category, result.title]
                      : ['home', result.category];

                  const displayTitle = result.displayTitle || result.title;

                  return (
                    <li key={result.id} className={styles.searchResult}>
                      <Link
                        href={result.slug}
                        className={styles.searchResultLink}
                      >
                        {displayTitle &&
                          (result.wrapInCode ? (
                            <code className={styles.searchResultCode}>
                              {displayTitle}
                            </code>
                          ) : (
                            <span className={styles.searchResultText}>
                              {displayTitle}
                            </span>
                          ))}
                        <SectionTitle path={sectionPath} />
                      </Link>
                    </li>
                  );
                })}
              </ul>
            )}
          </div>

          {Footer && (
            <div className={styles.searchResultsFooter}>
              <Footer />
            </div>
          )}
        </>
      )}
    </motion.div>
  );
};

export default SearchBar;
