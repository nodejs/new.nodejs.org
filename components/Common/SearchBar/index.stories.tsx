import SearchBar from './index';

import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof SearchBar>;
type Meta = MetaObj<typeof SearchBar>;

export const Default: Story = {
  args: {
    search() {
      return [
        {
          id: 'voluptate',
          slug: 'voluptate',
          title: 'Nisi deserunt excepteur',
          category: 'aliqua-sunt',
          displayTitle: 'Adipisicing magna irure elit velit.',
        },
        {
          id: 'ullamco',
          slug: 'ullamco',
          title: 'Lorem adipisicing ut',
          category: 'excepteur',
          displayTitle:
            'Aliqua voluptate aliqua non consectetur sunt consequat.',
        },
        {
          id: 'deserunt',
          slug: 'deserunt',
          title: 'Qui irure do irure',
          category: 'laborum',
          wrapInCode: true,
        },
      ];
    },
  },
};

export default { component: SearchBar } as Meta;
