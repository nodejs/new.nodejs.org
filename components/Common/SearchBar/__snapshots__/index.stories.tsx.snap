// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common/SearchBar Default smoke-test 1`] = `
<div class="SearchBar_searchBarContainer__Bo5vz"
     style="width: 0px; height: 3em;"
>
  <div class="SearchBar_searchInputContainer__Imb5N"
       role="presentation"
  >
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 24 24"
         class="SearchBar_searchIcon__nTZMm"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill="none"
            d="M0 0h24v24H0z"
      >
      </path>
      <path d="M19.3 16.9c.4-.7.7-1.5.7-2.4 0-2.5-2-4.5-4.5-4.5S11 12 11 14.5s2 4.5 4.5 4.5c.9 0 1.7-.3 2.4-.7l3.2 3.2 1.4-1.4-3.2-3.2zm-3.8.1c-1.4 0-2.5-1.1-2.5-2.5s1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5-1.1 2.5-2.5 2.5zM12 20v2C6.48 22 2 17.52 2 12S6.48 2 12 2c4.84 0 8.87 3.44 9.8 8h-2.07A8 8 0 0015 4.59V5c0 1.1-.9 2-2 2h-2v2c0 .55-.45 1-1 1H8v2h2v3H9l-4.79-4.79C4.08 10.79 4 11.38 4 12c0 4.41 3.59 8 8 8z">
      </path>
    </svg>
    <label class="SearchBar_searchLabel__d_ELb"
           for="searchInput"
    >
      <span>
        Search
      </span>
      <input autocomplete="off"
             class="SearchBar_searchInput__YjZjD"
             id="searchInput"
             name="query"
             type="text"
             value
      >
    </label>
  </div>
</div>
`;
